include/seqan3/range/view/kmer_hash.hpp

```diff
@@ -12,47 +12,647 @@
+         * \param skip Amount to increment.
+         * \attention This function is only avaible if `it_t` models std::RandomAccessIterator.
+         */
+        void hash_forward(difference_type const skip)
+        //!\cond
+            requires std::RandomAccessIterator<it_t>
+        //!\endcond
+        {
+            std::ranges::advance(text_left, skip);
+            hash_full();
+        }
+
+        /*!\brief Decrements iterator by 1.
+         * \attention This function is only avaible if `it_t` models std::BidirectionalIterator.
+         */
+        void hash_backward()
+        //!\cond
+            requires std::BidirectionalIterator<it_t>
+        //!\endcond
+        {
+            if (shape_.all())
+            {
+                hash_roll_backward();
+            }
+            else
+            {
+                std::ranges::advance(text_left,  -1);
+                hash_full();
+            }
+        }
+
+        /*!\brief Decrements iterator by `skip`.
+         * \param skip Amount to decrement.
+         * \attention This function is only avaible if `it_t` models std::BidirectionalIterator.
+         */
+        void hash_backward(difference_type const skip)
+        {
+            std::ranges::advance(text_left, -skip);
+            hash_full();
+        }
+
+        //!\brief Calculates a hash value by explicitly looking at each position.
+        void hash_full()
+        {
+            text_right = text_left;
+            hash_value = 0;
+
+            for (size_t i{0}; i < shape_.size() - 1u; ++i)
+            {
+                hash_value += shape_[i] * to_rank(*text_right);
+                hash_value *= shape_[i] ? sigma : 1;
+                std::ranges::advance(text_right, 1);
+            }
+        }
+
+        //!\brief Calculates the next hash value via rolling hash.
+        void hash_roll_forward()
+        {
+            hash_value -= to_rank(*(text_left)) * roll_factor;
+            hash_value += to_rank(*(text_right));
+            hash_value *= sigma;
+
+            std::ranges::advance(text_left,  1);
+            std::ranges::advance(text_right, 1);
+        }
+
+        /*!\brief Calculates the previous hash value via rolling hash.
+         * \attention This function is only avaible if `it_t` models std::BidirectionalIterator.
+         */
+        void hash_roll_backward()
+            //!\cond
+            requires std::BidirectionalIterator<it_t>
+            //!\endcond
+        {
+            std::ranges::advance(text_left,  -1);
+            std::ranges::advance(text_right, -1);
+
+            hash_value /= sigma;
+            hash_value -= to_rank(*(text_right));
+            hash_value += to_rank(*(text_left)) * roll_factor;
+        }
+    };
+
+/*!\name Associated types
+* \{
+*/
+//!\brief The iterator type of this view.
+using iterator          = shape_iterator<urng_t>;
+//!\brief The const_iterator type of this view.
+using const_iterator    = shape_iterator<urng_t const>;
+//!\}
```

Comment:
wrong indention

Reply:
Do you use these at all? You have return type auto below...

Reply:
```cpp
auto begin()
{
  return iterator{...};
}
```

I changed it to
```cpp
iterator begin()
{
  return {...};
}
```